Add ADF4382

chegbeli opened the pull request:
The ADF4382A is a high performance, ultralow jitter, Frac-N PLL
with integrated VCO ideally suited for LO generation for 5G applications
or data converter clock applications. The high performance
PLL has a figure of merit of -239 dBc/Hz, low 1/f Noise and
high PFD frequency of 625MHz in integer mode that can achieve
ultralow in-band noise and integrated jitter. The ADF4382A can
generate frequencies in a fundamental octave range of 11.5 GHz to
21 GHz, thereby eliminating the need for sub-harmonic filters. The
divide by 2 and 4 output dividers on the part allow frequencies to
be generated from 5.75GHz to 10.5GHz and 2.875GHz to 5.25GHz
respectively.

chegbeli commented:
V0 -> V1
Fixed DTB example missing parentheses.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+/* ADF4382_REG0 */
+#define ADF4382_ADDR_ASC_MSK			BIT(2)
+#define ADF4382_ADDR_ASC(x)			FIELD_PREP(ADF4382_ADDR_ASC_MSK, x)
```
drop all these parametrized definitions. i think it is preferred using FIELD_PREP/FIELD_GET inline.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+#define ADF4382_VCO_CAL_VTUNE			124
+#define ADF4382_VCO_CAL_ALC			250
+
+#define UA_TO_A					1000000
```
These should go in `units.h`, if not already there :)

nunojsa replied:
they should be there, yes

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	ret = regmap_write(st->regmap, 0x1A, var);
+	if (ret)
+		return ret;
+	var = (mod2_word >> 8) & ADF4382_MOD2WORD_MID_MSK;
```
instead of defining multiple GENMASK(7,0) for ADF4382_MOD2WORD_MID_MSK, ADF4382_MOD2WORD_MSB_MSK etc, i'd create some generic macros:
` #define ADF4382_WORD_LSB_MSK GENMASK(7, 0)`
` #define ADF4382_WORD_MID_MSK GENMASK(15, 8)`
` #define ADF4382_WORD_MSB_MSK GENMASK(23, 16)`

In this manner you drop multiple identical macro definitions and also get rid of explicit shifting by using FIELD_GET instead.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+static int adf4382_get_freq(struct adf4382_state *st, u64 *freq)
+{
+	*freq = st->freq;
```
we should obtain the frequency from hardware instead of cached values.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+		ret = adf4382_set_out_power(st, chan->channel, val);
+		break;
+	default:
+		ret = -EINVAL;
```
return directly.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	if (ret)
+		return ret;
+
+	mutex_lock(&st->lock);
```
i'd drop this mutex handling. it should go inside `adf4382_set_freq`. Moreover, `adf4382_set_out_power` doesn't seem to need locking.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	ret = device_property_read_u32(&st->spi->dev, "adi,ref-divider",
+				       &tmp);
+	if (ret || !tmp)
```
hmm, you treat "adi,ref-divider" property as optional, although you made it a required property in the dt bindings.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	ret = device_property_read_u32(&st->spi->dev, "adi,charge-pump-current",
+				       &tmp);
+	if (ret)
```
same here: property is optional in driver, while required in the dt bindings.

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+		return ret;
+
+	if (val != ADF4382_SCRATCHPAD_VAL) {
+		dev_err(&st->spi->dev, "Scratch pad test failed please check SPI connection");
```
dev_err_probe

amiclaus commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	ret = adf4382_init(st);
+	if (ret) {
+		dev_err(&spi->dev, "adf4382 init failed\n");
```
ditto: dev_err_probe.

amiclaus reviewed (commented):
some initial comments on my side. I'd suggest sending it upstream (although i noticed the datasheet not public yet, so the process might be harsher).

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf4382.yaml:
```diff
@@ -0,0 +1,142 @@
+      - minimum: 687500000
+      - maximum: 22000000000
+      - default: 2305000000
+    maxItems: 1
```
Not sure if the the above two properties belong in dts. The frequency one will be an hard sell upstream since I suppose that is something you can do already at runtime (changing the frequency). So you need to have a very good justification for this.

The  adi,bleed-word also seems like something we could/should have at runtime? Or is this word something dependent on the system/design and that we typically just want to program once?

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf4382.yaml:
```diff
@@ -0,0 +1,142 @@
+  adi,cmos-3v3:
+    description: |
+      Sets the SPI logic to 3.3V, by defautl it uses 1,8V.
+    type: boolean
```
Make sure the above two properties are not already present in:

https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/spi/spi-peripheral-props.yaml

nunojsa commented on Documentation/devicetree/bindings/iio/frequency/adi,adf4382.yaml:
```diff
@@ -0,0 +1,142 @@
+  - adi,ref-divider
+
+additionalProperties: false
+
```
The driver looks to be a clock provider but I'm not seeing any provider properties in the bindings...

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * adf4382 driver
```
This does not add much. Typically there's  a small subject (like the datasheet title). Look at other examples please

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	u64			vco_min;
+};
+
+//Charge pump current values expressed in uA
```
pleas don't use c++ comment style. In the kernel we do `/**/`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	7200,
+	8600,
+	10100
+};
```
nit: no need for a value per line...

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	{ 0x012, 0x00 },
+	{ 0x011, 0x00 },
+	{ 0x010, 0x50 },
+};
```
ditto

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	.reg_bits = 16,
+	.val_bits = 8,
+	.read_flag_mask = BIT(7),
+};
```
this is only used in the probe function so ideally you should reduce it's scope

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	return 0;
+}
+
+int adf4382_pll_fract_n_compute(struct adf4382_state *st, unsigned int pfd_freq_hz,
```
`static`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+int adf4382_frac2_compute(struct adf4382_state *st, u64 res,
+			  unsigned int pfd_freq_hz, u32 *frac2_word,
+			  u32 *mod2_word)
```
`static`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	if (ret)
+		return ret;
+
+	// dev->phase_adj = phase_ps;
```
drop unused code

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	/* Phase adjustment can only be done if bleed is active, and a bleed
+	 * constant needs to be added
+	 */
```
multiple lines should be like this:

```
/*
 * Phase adjustment can only be done if bleed is active, and a bleed
 * constant needs to be added
 */
```

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	if (!ch) {
+		tmp = FIELD_PREP(ADF4382_CLK1_OPWR_MSK, pwr);
+		return regmap_update_bits(st->regmap, 0x29, ADF4382_CLK1_OPWR_MSK,
+					  tmp);
```
just do the `FIELD_PREP()` in the function call directly...

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	 * values > 2^32 in order to support the entire frequency range
+	 * in Hz.
+	 */
+	_ADF4382_EXT_INFO("frequency", IIO_SHARED_BY_TYPE, ADF4382_FREQ),
```
Take a look at:

https://elixir.bootlin.com/linux/latest/source/include/linux/iio/types.h#L31

You can already do 64bit in the standard interface...

For the output power, I would double check in the [IIO ABI](https://elixir.bootlin.com/linux/latest/source/Documentation/ABI/testing/sysfs-bus-iio) to see if there's something that fits. Otherwise you'll have to justify the non standard attribute and you'll need an ABI file.

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+						       "adi,ref-doubler-enable");
+	st->cmos_3v3 = device_property_read_bool(&st->spi->dev, "adi,cmos-3v3");
+
+	st->clkin = devm_clk_get(&st->spi->dev, "ref_clk");
```
likely you want `devm_clk_get_enabled()`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	if (IS_ERR(st->clkin))
+		return PTR_ERR(st->clkin);
+
+	return 0;
```
return `PTR_ERR_OR_ZERO()`?

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	if (IS_ERR(clk))
+		return PTR_ERR(clk);
+
+	return of_clk_add_provider(of_node, of_clk_src_simple_get, clk);
```
`devm_of_clk_add_hw_provider()`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+	st = iio_priv(indio_dev);
+
+	indio_dev->info = &adf4382_info;
+	indio_dev->name = "adf4382";
```
you support two devices so ideally you would have the proper name when using adf4382a. Maybe have a chip_info struct only with `const char *name`

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	ret = clk_prepare_enable(st->clkin);
+	if (ret)
+		return ret;
```
you can drop the above with one of my previous comments

nunojsa commented on drivers/iio/frequency/adf4382.c:
```diff
@@ -0,0 +1,1537 @@
+
+	adf4382_setup_clk(st);
+	if (ret)
+		return ret;
```
The above is questionable. We do it for a couple of drivers and remember there was a reason for having both clock provider and an IIO userspace interface (like having different waveforms through IIO). But if the only reason is to have both userspace access top changing the frequency and having the clock exported as part of CCF at the same time, that will be more difficult to justify.

Anyways, one thing that should be protected is that if the channel is being exported through CCF, we should not allow to change it's frequency though the IIO interface as that could breaks potential consumers of the clock,

Anyways, let's see what upstream has to say about this.

nunojsa reviewed (changes requested):
First set of comments on my side.

The most tricky stuff is thje CCF abuse together with the IIO userspace interface. I know we do it in more drivers but... Anyways, as always, I'm expecting this driver to be upstreamed first (as I don't see any reason why we can't upstream it) so let's see what they have to say about this.

chegbeli commented:
V1->V2
- Fixed description.
- Remove param definitions.
- Using units.h, although I'm unsure if this shouldn't be improved.
- Replaced clock_scale with a fixed value, based on cpu architecture.
- Added enable and disable to clock ops.
- Implemented recalculating clock from registers.
- Changed the phase adjust functionality for higher accuracy.
- Added auto-align based on external pins to the phase adjust.
- Added displaying phase adjust value based on register value.
- Added debugfs fields for the bleed current values which are set.
  The bleed current changes based on phase adjust value.
- Added global initial output power level to DTS which will effect,
  both channels.

nunojsa commented:
@chegbeli long time for this one... Can you resolve the conflicts and follow up on this or should we close this PR?
